runtime/src/identity.rs

```diff
@@ -778,10 +839,115 @@ decl_event!(
+        ),
+
+        /// Authorization revoked or consumed. (auth_id, authorized_identity)
+        AuthorizationRemoved(u64, IdentityId),
     }
 );
 
 impl<T: Trait> Module<T> {
+    pub fn add_auth(
+        from_did: IdentityId,
+        target_did: IdentityId,
+        authorization_data: AuthorizationData,
+        expiry: Option<T::Moment>,
+    ) {
+        let new_nonce = Self::multi_purpose_nonce() + 1u64;
+        // Even if this transaction fails, nonce should be increased for added unpredictability
+        <MultiPurposeNonce>::put(&new_nonce);
+
+        let last_auth = Self::last_authorization(&target_did);
+
+        if last_auth > 0 {
+            //0 means no previous auth. 0 is the default value.
+            // Changing the last auth to point to new auth as next auth.
+            <Authorizations<T>>::mutate((target_did, last_auth), |last_authorization| {
+                last_authorization.next_authorization = new_nonce
+            });
+        }
+
+        let auth = Authorization {
+            authorization_data: authorization_data.clone(),
+            authorized_by: from_did,
+            expiry: expiry,
+            next_authorization: 0,
+            previous_authorization: last_auth,
+        };
+
+        <LastAuthorization>::insert(&target_did, new_nonce);
+        <Authorizations<T>>::insert((target_did, new_nonce), auth);
+
+        Self::deposit_event(RawEvent::NewAuthorization(
+            new_nonce,
+            from_did,
+            target_did,
+            authorization_data,
+            expiry,
+        ));
+    }
+
+    /// Remove any authorization. No questions asked.
+    /// NB: Please do all the required checks before calling this function.
+    pub fn remove_auth(target_did: IdentityId, auth_id: u64, next_auth: u64, previous_auth: u64) {
+        if next_auth != 0 {
+            // update next auth's previous auth to point to previous auth of this auth
+            <Authorizations<T>>::mutate((target_did, next_auth), |next_auth| {
+                next_auth.previous_authorization = previous_auth
+            });
+        } else {
+            // this was the last auth. update last auth to be previous auth.
+            <LastAuthorization>::insert(&target_did, previous_auth);
+        }
+        if previous_auth != 0 {
+            // update previous auth's next auth to point to next auth of this auth
+            <Authorizations<T>>::mutate((target_did, previous_auth), |prev_auth| {
+                prev_auth.next_authorization = next_auth
+            });
+        }
+        <Authorizations<T>>::remove((target_did, auth_id));
+        Self::deposit_event(RawEvent::AuthorizationRemoved(auth_id, target_did));
+    }
+
+    /// Consumes an authorization.
+    /// Checks if the auth has not expired and the caller is authorized to consume this auth.
+    pub fn consume_auth(
+        from_did: IdentityId,
+        target_did: IdentityId,
+        auth_id: u64,
+    ) -> AuthorizationStatus {
```

Comment:
Suggestion: Replace returned type by `Result<(), AuthorizationError>`.
In that way, it follows the common design in Rust, and it will facilitate the error process. i.e:
```
match consume_auth( from, target, auth_id) {
   Ok(_) => { ... },
   Err(error) =>  print( "{:?}", error)
}
```
or 
```
if let Err(error) = consume_auth( from, target, auth_id) {
   print( "{:?}", error);
   // Or early return
}
```
And impl `Display` for `AuthorizationError` (just one place).